In [ ]:
import numpy as np
import pandas as pd

heart_df = pd.read_csv('heart_failure_dataset.csv')
####1.1####
print("**a**")
print("shape:")
print(heart_df.shape) 
print("axes:")
print(heart_df.axes)
print("dtypes:")
print(heart_df.dtypes)
print("\n")
print("**b**")
print(heart_df.head(10))


**a**
shape:
(299, 13)
axes:
[RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]
dtypes:
age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


**b**
    age  anaemia  creatinine_phosphokinase  ...  smoking  time  DEATH_EVENT
0  75.0        0                       582  ...        0     4            1
1  55.0        0                     

\*\*c\*\* 
> Each row of heart_df dataframe is termed an "instance" in machine learning parlance.

\*\*d\*\*
> This machine learning problem is a supervised learning problem, because we need to tell the algorithm whether the patient died or not after the data was observed.

\*\*e\*\*
> Independent Variables: (12) age, anaemia, creatinine_phosphokinase, diabetes, ejection_fraction, high_blood_pressure, platelets, serum_creatinine, serum_sodium, sex, smoking, time

\*\*f\*\*
> Dependent Variables: (1) DEATH_EVENT

\*\*g\*\*
> Categorical Variables: anaemia, diabetes, high_blood_pressure, sex, smoking, DEATH_EVENT

\*\*h\*\*
> The categortical variables follow label encoding

\*\*i\*\*
> 203 patients did not die during follow-up period
96 patients died during follow-up period

\*\*j\*\*
> 105 female, 194 male patients

\*\*k\*\*
> 203 non-smokers, 96 smokers










In [ ]:
####1.2####
heart_df.corr()

age   anaemia  ...      time  DEATH_EVENT
age                       1.000000  0.088006  ... -0.224068     0.253729
anaemia                   0.088006  1.000000  ... -0.141414     0.066270
creatinine_phosphokinase -0.081584 -0.190741  ... -0.009346     0.062728
diabetes                 -0.101012 -0.012729  ...  0.033726    -0.001943
ejection_fraction         0.060098  0.031557  ...  0.041729    -0.268603
high_blood_pressure       0.093289  0.038182  ... -0.196439     0.079351
platelets                -0.052354 -0.043786  ...  0.010514    -0.049139
serum_creatinine          0.159187  0.052174  ... -0.149315     0.294278
serum_sodium             -0.045966  0.041882  ...  0.087640    -0.195204
sex                       0.065430 -0.094769  ... -0.015608    -0.004316
smoking                   0.018668 -0.107290  ... -0.022839    -0.012623
time                     -0.224068 -0.141414  ...  1.000000    -0.526964
DEATH_EVENT               0.253729  0.066270  ... -0.526964     1.000000

[13 rows x 13 columns]

\*\*a\*\*
> age (0.253729), and serum_creatinine (0.294278)

\*\*b\*\*
> time (-0.526964), and ejection_fraction (-0.268603)

\*\*c\*\*
> The older you get the more likely you are to have a heart failure event.

\*\*d\*\*
> The time period of followup does not have anything to do with the death event, and ejection_fraction is not a telltale sign of heart failure.







In [ ]:
####1.3####
# define a function to check if a column that should be binary
# contains only binary values
def checkIfBinary(s):
  for i in range(len(s)):
    if(s[i] < 0 or s[i] > 1): 
      s[i] = float("NaN")

# any values in age series <= 0 converted to NaN
for i in range(len(heart_df.age)):
  if(heart_df.age[i] <= 0): 
    heart_df.age[i] = float("NaN")

# any non-binary values in anaemia series converted to NaN
checkIfBinary(heart_df.anaemia)

# any values in creatinine_phosphokinase <= 0 converted to NaN
for i in range(len(heart_df.creatinine_phosphokinase)):
  if(heart_df.creatinine_phosphokinase[i] <= 0): 
    heart_df.creatinine_phosphokinase[i] = float("NaN")

# any non-binary values in diabetes series converted to NaN
checkIfBinary(heart_df.diabetes)

# any values in 100 < ejection_fraction < 0 converted to NaN
for i in range(len(heart_df.ejection_fraction)):
  if(heart_df.ejection_fraction[i] < 0 or heart_df.ejection_fraction[i] > 100): 
    heart_df.ejection_fraction[i] = float("NaN")

# any non-binary values in diabetes series converted to NaN
checkIfBinary(heart_df.high_blood_pressure)  

# any values in platelets <= 0 converted to NaN
for i in range(len(heart_df.platelets)):
  if(heart_df.platelets[i] <= 0): 
    heart_df.platelets[i] = float("NaN")

# any values in serum_creatinine <= 0 converted to NaN
for i in range(len(heart_df.serum_creatinine)):
  if(heart_df.serum_creatinine[i] <= 0): 
    heart_df.serum_creatinine[i] = float("NaN")

# any values in serum_sodium <= 0 converted to NaN
for i in range(len(heart_df.serum_sodium)):
  if(heart_df.serum_sodium[i] <= 0): 
    heart_df.serum_sodium[i] = float("NaN")

# any non-binary values in sex series converted to NaN
checkIfBinary(heart_df.sex) 

# any non-binary values in smoking series converted to NaN
checkIfBinary(heart_df.smoking) 

# any values in time <= 0 converted to NaN
for i in range(len(heart_df.time)):
  if(heart_df.time[i] <= 0): 
    heart_df.time[i] = float("NaN")

# any non-binary values in DEATH_EVENT series converted to NaN
checkIfBinary(heart_df.DEATH_EVENT) 

heart_df.to_csv('heart_failure_dataset_NaNs.csv', index=False)

In [ ]:
####1.4####
# convert columns to store floats, needed to read zero mean and unit variance
x = heart_df.astype('float64')

# store mean and standard deviation of each series before iterating through series
mean = x.age.mean()
std = x.age.std()

# iterate through series and normalize values
for i in range(len(x.age)):
  if x.age[i] != float("NaN"): 
    x.age[i] = (float(x.age[i]) - mean) / std

mean = x.creatinine_phosphokinase.mean()
std = x.creatinine_phosphokinase.std()
for i in range(len(x.creatinine_phosphokinase)):
  if x.creatinine_phosphokinase[i] != float("NaN"): 
    x.creatinine_phosphokinase[i] = (float(x.creatinine_phosphokinase[i]) - mean) / std

mean = x.ejection_fraction.mean()
std = x.ejection_fraction.std()
for i in range(len(x.ejection_fraction)):
  if x.ejection_fraction[i] != float("NaN"): 
    x.ejection_fraction[i] = (float(x.ejection_fraction[i]) - mean) / std

mean = x.platelets.mean()
std = x.platelets.std()
for i in range(len(x.platelets)):
  if x.platelets[i] != float("NaN"): 
    x.platelets[i] = (float(x.platelets[i]) - mean) / std

mean = x.serum_creatinine.mean()
std = x.serum_creatinine.std()
for i in range(len(x.serum_creatinine)):
  if x.serum_creatinine[i] != float("NaN"): 
    x.serum_creatinine[i] = (float(x.serum_creatinine[i]) - mean) / std

mean = x.serum_sodium.mean()
std = x.serum_sodium.std()
for i in range(len(x.serum_sodium)):
  if x.serum_sodium[i] != float("NaN"): 
    x.serum_sodium[i] = (float(x.serum_sodium[i]) - mean) / std

mean = x.time.mean()
std = x.time.std()
for i in range(len(x.time)):
  if x.time[i] != float("NaN"): 
    x.time[i] = (float(x.time[i]) - mean) / std

print(x.head(20))
  

         age  anaemia  creatinine_phosphokinase  ...  smoking      time  DEATH_EVENT
0   1.190949      0.0                  0.000165  ...      0.0 -1.626775          1.0
1  -0.490457      0.0                  7.502063  ...      0.0 -1.601007          1.0
2   0.350246      0.0                 -0.449186  ...      1.0 -1.588122          1.0
3  -0.910808      1.0                 -0.485257  ...      0.0 -1.588122          1.0
4   0.350246      1.0                 -0.434757  ...      0.0 -1.575238          1.0
5   2.452003      1.0                 -0.551217  ...      1.0 -1.575238          1.0
6   1.190949      1.0                 -0.346124  ...      0.0 -1.549470          1.0
7  -0.070106      1.0                 -0.275011  ...      1.0 -1.549470          1.0
8   0.350246      0.0                 -0.437849  ...      0.0 -1.549470          1.0
9   1.611300      1.0                 -0.472890  ...      1.0 -1.549470          1.0
10  1.190949      1.0                 -0.516176  ...      1.0 -1.

In [ ]:
####1.5####
# pop death event column from dataframe
death_event = x.pop('DEATH_EVENT')
# create zeroed len(x.columns) x 2 array of type int64
death = np.zeros(((x.last_valid_index() + 1), 2), dtype='int64')
# iterate and one-hot-encode values of death_event into death array
for i in range(len(death_event)):
  if death_event[i] == 0:
    death[i][0] = 1
    death[i][1] = 0
  else:
    death[i][0] = 0
    death[i][1] = 1
# insert death columns into dataframe
x.insert(loc = len(x.columns), column='NO_DEATH', value=death[:, 0])
x.insert(loc = len(x.columns), column='DEATH', value=death[:, 1])
# save csv file
x.to_csv('heart_failure_dataset_processed.csv', index=False)


In [ ]:
###2.1###
diabetes_df = pd.read_csv('diabetes_dataset.csv')
print("**a**")
print("shape:")
print(diabetes_df.shape) 
print("axes:")
print(diabetes_df.axes)
print("dtypes:")
print(diabetes_df.dtypes)
print("\n")
print("**b**")
print(diabetes_df.head(10))

**a**
shape:
(768, 9)
axes:
[RangeIndex(start=0, stop=768, step=1), Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')]
dtypes:
Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


**b**
   Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21      

\*\*c\*\*
> This machine learning problem is a supervised learning problem, because we need to tell the algorithm whether the patient died or not after the data was observed.

\*\*d\*\*
> Independent Variables: (8) Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age

\*\*e\*\*
> Dependent Variables: (1) Outcome

\*\*f\*\*
> Categorical Variables: Outcome

\*\*g\*\*
> The categorical variables follow label encoding

\*\*h\*\*
> Young Adults (21-40): 574

> Middle-Aged Adults (41-60): 167

> Old-Aged Adults (61+): 27

In [ ]:
young = 0
middle = 0
old = 0
for i in range(len(diabetes_df.Age)):
    if diabetes_df.Age[i] >= 21 and diabetes_df.Age[i] <= 40:
      young = young + 1
    elif diabetes_df.Age[i] >= 41 and diabetes_df.Age[i] <= 60:
      middle = middle + 1
    elif diabetes_df.Age[i] >= 61:
      old = old + 1
    i = i + 1
print(young)
print(middle)
print(old)


574
167
27


In [ ]:
####2.2####
# any values in pregnancies series < 0 converted to NaN
for i in range(len(diabetes_df.Pregnancies)):
  if(diabetes_df.Pregnancies[i] < 0): 
    diabetes_df.Pregnancies[i] = float("NaN")
# any values in Glucose series <= 0 converted to NaN   
for i in range(len(diabetes_df.Glucose)):
  if(diabetes_df.Glucose[i] <= 0): 
    diabetes_df.Glucose[i] = float("NaN")
# any values in BloodPressure series <= 0 converted to NaN   
for i in range(len(diabetes_df.BloodPressure)):
  if(diabetes_df.BloodPressure[i] <= 0): 
    diabetes_df.BloodPressure[i] = float("NaN")
# any values in SkinThickness series <= 0 converted to NaN   
for i in range(len(diabetes_df.SkinThickness)):
  if(diabetes_df.SkinThickness[i] <= 0): 
    diabetes_df.SkinThickness[i] = float("NaN")
# any values in Insulin series <= 0 converted to NaN
for i in range(len(diabetes_df.Insulin)):
  if(diabetes_df.Insulin[i] <= 0): 
    diabetes_df.Insulin[i] = float("NaN")
# any values in BMI series <= 0 converted to NaN
for i in range(len(diabetes_df.BMI)):
  if(diabetes_df.BMI[i] <= 0): 
    diabetes_df.BMI[i] = float("NaN")
# any values in DiabetesPedigreeFunction series <= 0 converted to NaN
for i in range(len(diabetes_df.DiabetesPedigreeFunction)):
  if(diabetes_df.DiabetesPedigreeFunction[i] <= 0): 
    diabetes_df.DiabetesPedigreeFunction[i] = float("NaN")
# any values in Age series <= 0 converted to NaN
for i in range(len(diabetes_df.Age)):
  if(diabetes_df.Age[i] <= 0): 
    diabetes_df.Age[i] = float("NaN")
# any non-binary values in Outcome series converted to NaN
checkIfBinary(diabetes_df.Outcome)

print(diabetes_df.head(20))

    Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0             6    148.0           72.0  ...                     0.627   50        1
1             1     85.0           66.0  ...                     0.351   31        0
2             8    183.0           64.0  ...                     0.672   32        1
3             1     89.0           66.0  ...                     0.167   21        0
4             0    137.0           40.0  ...                     2.288   33        1
5             5    116.0           74.0  ...                     0.201   30        0
6             3     78.0           50.0  ...                     0.248   26        1
7            10    115.0            NaN  ...                     0.134   29        0
8             2    197.0           70.0  ...                     0.158   53        1
9             8    125.0           96.0  ...                     0.232   54        1
10            4    110.0           92.0  ...                     

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]
/usr/local/lib/pyt

\#\#\#\#2.2\#\#\#\#
People cannot be pregnant less than 0 times
Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, and Age should all be > 0 
Outcome column should only contain 0s and 1s, any other input is invalid.

In [ ]:
####2.3####
# convert columns to store floats, needed to read zero mean and unit variance
x = diabetes_df.astype('float64')

# store mean and standard deviation of each series before iterating through series
mean = x.Pregnancies.mean()
std = x.Pregnancies.std()

# iterate through series and normalize values
for i in range(len(x.Pregnancies)):
  if x.Pregnancies[i] != float("NaN"): 
    x.Pregnancies[i] = (float(x.Pregnancies[i]) - mean) / std

mean = x.Glucose.mean()
std = x.Glucose.std()
for i in range(len(x.Glucose)):
  if x.Glucose[i] != float("NaN"): 
    x.Glucose[i] = (float(x.Glucose[i]) - mean) / std

mean = x.BloodPressure.mean()
std = x.BloodPressure.std()
for i in range(len(x.BloodPressure)):
  if x.BloodPressure[i] != float("NaN"): 
    x.BloodPressure[i] = (float(x.BloodPressure[i]) - mean) / std

mean = x.SkinThickness.mean()
std = x.SkinThickness.std()
for i in range(len(x.SkinThickness)):
  if x.SkinThickness[i] != float("NaN"): 
    x.SkinThickness[i] = (float(x.SkinThickness[i]) - mean) / std

mean = x.Insulin.mean()
std = x.Insulin.std()
for i in range(len(x.Insulin)):
  if x.Insulin[i] != float("NaN"): 
    x.Insulin[i] = (float(x.Insulin[i]) - mean) / std

mean = x.BMI.mean()
std = x.BMI.std()
for i in range(len(x.BMI)):
  if x.BMI[i] != float("NaN"): 
    x.BMI[i] = (float(x.BMI[i]) - mean) / std

mean = x.DiabetesPedigreeFunction.mean()
std = x.DiabetesPedigreeFunction.std()
for i in range(len(x.DiabetesPedigreeFunction)):
  if x.DiabetesPedigreeFunction[i] != float("NaN"): 
    x.DiabetesPedigreeFunction[i] = (float(x.DiabetesPedigreeFunction[i]) - mean) / std

mean = x.Age.mean()
std = x.Age.std()
for i in range(len(x.Age)):
  if x.Age[i] != float("NaN"): 
    x.Age[i] = (float(x.Age[i]) - mean) / std

In [ ]:
####2.4####
# pop outcome column from dataframe
outcome = x.pop('Outcome')
# create zeroed 768 x 2 array of type int64
diabetes = np.zeros(((x.last_valid_index() + 1), 2), dtype='int64')
# iterate and one-hot-encode values of outcome into diabetes array
for i in range(len(outcome)):
  if outcome[i] == 0:
    diabetes[i][0] = 1
    diabetes[i][1] = 0
  else:
    diabetes[i][0] = 0
    diabetes[i][1] = 1
# insert diabetes columns into dataframe
x.insert(loc = len(x.columns), column='NO_DIABETES', value=diabetes[:, 0])
x.insert(loc = len(x.columns), column='DIABETES', value=diabetes[:, 1])

In [ ]:
####2.5####
# initialize list of values we will use to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.Pregnancies)):
  if not (np.isnan(x.Pregnancies[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.Pregnancies[i])
    else:
      oneOutcomeVals.append(x.Pregnancies[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.Pregnancies)):
  if np.isnan(x.Pregnancies[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.Pregnancies[i] = zeroMedian
    else:
      x.Pregnancies[i] = oneMedian

# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.Glucose)):
  if not (np.isnan(x.Glucose[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.Glucose[i])
    else:
      oneOutcomeVals.append(x.Glucose[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.Glucose)):
  if np.isnan(x.Glucose[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.Glucose[i] = zeroMedian
    else:
      x.Glucose[i] = oneMedian
  
# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.BloodPressure)):
  if not (np.isnan(x.BloodPressure[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.BloodPressure[i])
    else:
      oneOutcomeVals.append(x.BloodPressure[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.BloodPressure)):
  if np.isnan(x.BloodPressure[i]): # if NaN value, replace with median from its respective outcome
    if y.NO_DIABETES[i] == 1:
      x.BloodPressure[i] = zeroMedian
    else:
      x.BloodPressure[i] = oneMedian

# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.SkinThickness)):
  if not (np.isnan(x.SkinThickness[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.SkinThickness[i])
    else:
      oneOutcomeVals.append(x.SkinThickness[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.SkinThickness)):
  if np.isnan(x.SkinThickness[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.SkinThickness[i] = zeroMedian
    else:
      x.SkinThickness[i] = oneMedian

# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.Insulin)):
  if not (np.isnan(x.Insulin[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.Insulin[i])
    else:
      oneOutcomeVals.append(x.Insulin[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.Insulin)):
  if np.isnan(x.Insulin[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.Insulin[i] = zeroMedian
    else:
      x.Insulin[i] = oneMedian

# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.BMI)):
  if not (np.isnan(x.BMI[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.BMI[i])
    else:
      oneOutcomeVals.append(x.BMI[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.BMI)):
  if np.isnan(x.BMI[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.BMI[i] = zeroMedian
    else:
      x.BMI[i] = oneMedian
    
# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.DiabetesPedigreeFunction)):
  if not (np.isnan(x.DiabetesPedigreeFunction[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.DiabetesPedigreeFunction[i])
    else:
      oneOutcomeVals.append(x.DiabetesPedigreeFunction[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.DiabetesPedigreeFunction)):
  if np.isnan(x.DiabetesPedigreeFunction[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.DiabetesPedigreeFunction[i] = zeroMedian
    else:
      x.DiabetesPedigreeFunction[i] = oneMedian

# reset list of values used to calculate median
zeroOutcomeVals = list()
oneOutcomeVals = list()
# iterate through column
for i in range(len(x.Age)):
  if not (np.isnan(x.Age[i])): # if not NaN value, add to either zero outcome list or one outcome list
    if x.NO_DIABETES[i] == 1:
      zeroOutcomeVals.append(x.Age[i])
    else:
      oneOutcomeVals.append(x.Age[i])
# calculate medians for both zero outcome instances and one outcome instances
zeroMedian = np.median(zeroOutcomeVals)
oneMedian = np.median(oneOutcomeVals)
# iterate through column
for i in range(len(x.Age)):
  if np.isnan(x.Age[i]): # if NaN value, replace with median from its respective outcome
    if x.NO_DIABETES[i] == 1:
      x.Age[i] = zeroMedian
    else:
      x.Age[i] = oneMedian

# output to csv file

x.to_csv('diabetes_dataset_processed.csv', index=False)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:42: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:61: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:63: SettingWithCopyWarning: 
A value is trying to b